### ISSUE TITLE
ConnBuilder loop_over_targets() does not work as intended

### ISSUE THREAD

**Issue by heplesser** [nest/nest-simulator#3106]
Following Ippen et al (2017, Sec 3.2), ConnBuilders should loop over local nodes instead of over targets when the number of local nodes is smaller than the target population. Unfortunately, ever since 20b8f338231e142ab88971a8423b130caaafc2c7, the `loop_over_targets()` method looks like this:

https://github.com/nest/nest-simulator/blob/8e85268ca512e0f509240fb5ef06f62a4e61a67e/nestkernel/conn_builder.cpp#L404C1-L410C1

Here, `kernel().node_manager.size()` returns the size of the entire network, so the condition will always be true unless `targets` contains the entire network. This is not what is intended.

I have not checked fully, but I believe we should compare `targets.size()` to the number of nodes on the pertaining thread.

**Comment by heplesser** [comment]
@note #3132 will **not** address this issue, that has to wait for a later PR. For an Ansatz to a solution is in 33fa658ac53ab2d65f8e3aceeedf5e0f8ffc671c.

**Comment by jessica-mitchell** [comment]
@heplesser has this issue had any developments? is it still relevant?

**Comment by heplesser** [comment]
Yes, this is still relevant, may be a topic for a future hackathon.
